File: go/allocation_realistic.go
// Benchmark 3b: Realistic allocation (prevents optimization)
// Run: go run allocation_realistic.go
// Compare with: go run -gcflags="-m" allocation_realistic.go 2>&1 | grep escape

package main

import (
	"fmt"
	"time"
	"unsafe"
)

type Point struct {
	X, Y int
	Data [10]int  // Make it bigger to prevent optimizations
}

// Global to prevent optimizer from eliminating allocations
var gSum int64

// Benchmark heap allocation (store pointers in slice)
func benchmarkHeapRealistic(n int) time.Duration {
	points := make([]*Point, 0, n)
	
	start := time.Now()
	
	// Allocate (escapes to heap)
	for i := 0; i < n; i++ {
		p := &Point{}
		p.X = i
		p.Y = i
		points = append(points, p)  // Store pointer (forces escape)
	}
	
	allocEnd := time.Now()
	
	// Use the data (prevents dead code elimination)
	sum := int64(0)
	for _, p := range points {
		sum += int64(p.X + p.Y)
	}
	gSum = sum
	
	return allocEnd.Sub(start)
}

// Benchmark stack allocation (values in slice)
func benchmarkStackRealistic(n int) time.Duration {
	points := make([]Point, 0, n)
	
	start := time.Now()
	
	// Allocate (stays in slice's contiguous memory)
	for i := 0; i < n; i++ {
		p := Point{}
		p.X = i
		p.Y = i
		points = append(points, p)  // Store value
	}
	
	allocEnd := time.Now()
	
	// Use the data (prevents dead code elimination)
	sum := int64(0)
	for _, p := range points {
		sum += int64(p.X + p.Y)
	}
	gSum = sum
	
	return allocEnd.Sub(start)
}

func main() {
	const n = 1000000  // 1 million allocations
	
	fmt.Println("Benchmarking Go realistic allocation patterns")
	fmt.Printf("Allocations: %d\n", n)
	fmt.Printf("Object size: %d bytes\n\n", int(unsafe.Sizeof(Point{})))
	
	// Warm up
	benchmarkHeapRealistic(1000)
	benchmarkStackRealistic(1000)
	
	// Benchmark heap allocation
	heapTime := benchmarkHeapRealistic(n)
	heapMicros := heapTime.Microseconds()
	heapPerAlloc := heapTime.Nanoseconds() / int64(n)
	
	fmt.Println("Heap allocation (pointer slice):")
	fmt.Printf("  Total time: %.2f ms\n", float64(heapMicros)/1000.0)
	fmt.Printf("  Time per allocation: %d ns\n\n", heapPerAlloc)
	
	// Benchmark stack allocation
	stackTime := benchmarkStackRealistic(n)
	stackMicros := stackTime.Microseconds()
	stackPerAlloc := stackTime.Nanoseconds() / int64(n)
	
	fmt.Println("Value slice (contiguous storage):")
	fmt.Printf("  Total time: %.2f ms\n", float64(stackMicros)/1000.0)
	fmt.Printf("  Time per allocation: %d ns\n\n", stackPerAlloc)
	
	// Calculate speedup
	speedup := float64(heapTime) / float64(stackTime)
	fmt.Printf("Speedup: %.2fx faster for value storage\n", speedup)
	fmt.Println("\nNote: This measures allocation + initialization + append.")
	fmt.Println("Heap requires malloc per object, value slice grows contiguously.")
}

File: cpp/allocation_realistic.cpp
// Benchmark 3b: Realistic allocation (prevents optimization)
// Compile: g++ -O2 -std=c++17 -o allocation_realistic allocation_realistic.cpp
// Note: Using -O2 instead of -O3 to reduce aggressive optimization

#include <iostream>
#include <vector>
#include <chrono>
#include <random>

struct Point {
    int x, y;
    int data[10];  // Make it bigger to prevent optimizations
};

// Global to prevent optimizer from eliminating allocations
volatile long long g_sum = 0;

// Benchmark heap allocation (realistic: store pointers, use later)
long long benchmark_heap_realistic(size_t n) {
    std::vector<Point*> points;
    points.reserve(n);
    
    auto start = std::chrono::high_resolution_clock::now();
    
    // Allocate
    for (size_t i = 0; i < n; i++) {
        Point* p = new Point{};
        p->x = i;
        p->y = i;
        points.push_back(p);  // Store pointer (escapes)
    }
    
    auto alloc_end = std::chrono::high_resolution_clock::now();
    
    // Use the data (prevents dead code elimination)
    long long sum = 0;
    for (auto* p : points) {
        sum += p->x + p->y;
    }
    g_sum = sum;
    
    // Cleanup
    for (auto* p : points) {
        delete p;
    }
    
    return std::chrono::duration_cast<std::chrono::nanoseconds>(alloc_end - start).count();
}

// Benchmark stack allocation (realistic: values in vector)
long long benchmark_stack_realistic(size_t n) {
    std::vector<Point> points;
    points.reserve(n);
    
    auto start = std::chrono::high_resolution_clock::now();
    
    // Allocate (stored in vector's contiguous memory)
    for (size_t i = 0; i < n; i++) {
        Point p{};
        p.x = i;
        p.y = i;
        points.push_back(p);
    }
    
    auto alloc_end = std::chrono::high_resolution_clock::now();
    
    // Use the data (prevents dead code elimination)
    long long sum = 0;
    for (const auto& p : points) {
        sum += p.x + p.y;
    }
    g_sum = sum;
    
    return std::chrono::duration_cast<std::chrono::nanoseconds>(alloc_end - start).count();
}

int main() {
    const size_t n = 1000000;  // 1 million allocations
    
    std::cout << "Benchmarking realistic allocation patterns\n";
    std::cout << "Allocations: " << n << "\n";
    std::cout << "Object size: " << sizeof(Point) << " bytes\n\n";
    
    // Warm up
    benchmark_heap_realistic(1000);
    benchmark_stack_realistic(1000);
    
    // Benchmark heap allocation
    auto heap_time = benchmark_heap_realistic(n);
    auto heap_ms = heap_time / 1000000.0;
    auto heap_per_alloc = heap_time / n;
    
    std::cout << "Heap allocation (new + store pointer):\n";
    std::cout << "  Total time: " << heap_ms << " ms\n";
    std::cout << "  Time per allocation: " << heap_per_alloc << " ns\n\n";
    
    // Benchmark stack allocation
    auto stack_time = benchmark_stack_realistic(n);
    auto stack_ms = stack_time / 1000000.0;
    auto stack_per_alloc = stack_time / n;
    
    std::cout << "Stack allocation (push_back value):\n";
    std::cout << "  Total time: " << stack_ms << " ms\n";
    std::cout << "  Time per allocation: " << stack_per_alloc << " ns\n\n";
    
    // Calculate speedup
    double speedup = static_cast<double>(heap_time) / stack_time;
    std::cout << "Speedup: " << speedup << "x faster for stack-based storage\n";
    std::cout << "\nNote: This measures allocation + initialization + storage.\n";
    std::cout << "Heap requires malloc, stack requires vector growth (amortized).\n";
    
    return 0;
}
